overlord/snapstate/snapstate.go

```diff
@@ -134,7 +137,15 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		stop := st.NewTask("stop-snap-services", fmt.Sprintf(i18n.G("Stop snap %q services"), snapsup.Name()))
 		addTask(stop)
 		prev = stop
+	}
+
+	if runRefreshHooks {
+		preRefreshHook := SetupPreRefreshHook(st, snapsup.Name())
```

Comment:
Can the refresh hook restart the services because it runs after they are stopped and can poke their sockets and what not?

Reply:
Yes, it can, there is nothing that prevents it currently. I see how this can be an issue that leads to a failure if service is started for the old revision of the snap in pre-refresh. We can easily prevent this and error out in snapctl based on hook name, if that's what we want.

Reply:
Why are we running this after the services are stopped?  Part of the reason of running this before it stops is precisely so that we have the data and the old environment at hand. People will need the service running if we intend them to be able to dump data before the refresh, for example.